### PR TITLE
Improve sbt/gh actions infrastructure

### PR DESCRIPTION
**Pull request by armanbilge**
WIP more coming :)

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -5,11 +5,16 @@ lazy val root = project.in(file(".")).
 
 name := "Scala.js DOM"
 
-crossScalaVersions in ThisBuild := {
-  if (scalaJSVersion.startsWith("1.")) Seq("2.12.10", "2.11.12", "2.13.1")
-  else Seq("2.12.10", "2.11.12", "2.10.7", "2.13.1")
+lazy val scala210 = "2.10.7"
+lazy val scala211 = "2.11.12"
+lazy val scala212 = "2.12.10"
+lazy val scala213 = "2.13.1"
```
Comment (armanbilge) [inline build.sbt:11]: Might be good to update some of these, I think @sjrd likes to keep them in sync with their respective Scala.js version.

File: build.sbt [hunk 2]
```diff
@@ -5,14 +5,71 @@ lazy val root = project.in(file(".")).
...
+  publish = publish.copy(steps = publish.steps.map {
+    case step: WorkflowStep.Use if step.name.exists(_.startsWith("Download target directories")) =>
+      step.copy(params = step.params.mapValues(_ + "-${{ matrix.scalajs }}"))
+        .copy(cond = if (step.name.exists(_.contains(scala210))) Some(s"matrix.scalajs == '${scalaJS06}'") else None)
+    case step: WorkflowStep.Run if step.name.exists(_.contains(scala210)) =>
+      step.copy(cond = Some(s"matrix.scalajs == '${scalaJS06}'"))
+    case step => step
+  })
+  jobs.updated(1, publish)
+}
```
Comment (armanbilge) [inline build.sbt:55]: This whole mess is to cross on 0.6 and 1.0, I look forward to deleting it when we merge into master.

Comment (japgolly) [reply build.sbt:56]: Yeah 0.6 is such a pain this late in the game

**Comment by japgolly** [comment]
@armanbilge Just to confirm, this is ready for review and even merge and I believe that because you've assigned me as a reviewer. Is that right, and is that the process we'll follow to know? Sounds like a good idea to me

**Comment by armanbilge** [comment]
👍 yup that's the process: draft PR is WIP otherwise ready for review and merge with your approval. BTW this will badly conflict with the amazing work you've done in #471, I'm happy for you to merge first and adjust here accordingly.

**Comment by japgolly** [comment]
Also omg! The gh actions yamls have become monsters!! Plus what used to be making a small change to the yaml is now a bigger change in sbt. Sell me again on why this is better?

**Comment by japgolly** [comment]
> BTW this will badly conflict with the amazing work you've done in #471, I'm happy for you to merge first and adjust here accordingly.

hehe thanks and it's all good. You got in first so I'll update mine if this is merged first

**Comment by armanbilge** [comment]
> Sell me again on why this is better?

Lol!! Yeah this one is a _really_ bad example because of the SJS 0.6/1 cross stuff. It will get better when we can remove that.

The idea is that `sbt` is the authority on all the Scala versions and any other important settings (projects that need to be cached, etc. etc.). So we don't have to remember to manually keep them in sync with the yaml.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -1,47 +1,173 @@
...
+name: Continuous Integration
 
 on:
   pull_request:
+    branches: ['**']
   push:
-    tags-ignore:
-      - v*
+    branches: ['**']
+    tags: [v*]
```
Comment (japgolly) [inline .github/workflows/ci.yml:15]: This is the inverse of what we want :) Currently it's all tags except `v*`

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -1,47 +1,173 @@
...
+      - name: Upload target directories
+        uses: actions/upload-artifact@v2
+        with:
+          path: targets.tar
+          name: target-${{ matrix.os }}-${{ matrix.scala }}-${{ matrix.java }}-${{ matrix.scalajs }}
+
+  publish:
+    name: Publish Artifacts
+    needs: [build]
+    if: github.event_name != 'pull_request' && (startsWith(github.ref, 'refs/tags/v'))
```
Comment (armanbilge) [inline .github/workflows/ci.yml:93]: Actually it's fine, this is the where the check is whether to publish or not.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (armanbilge) [reply .github/workflows/ci.yml:15]: See https://github.com/scala-js/scala-js-dom/pull/466/files#r687287273, I think it's fine like this?

**Comment by japgolly** [comment]
> The idea is that sbt is the authority on all the Scala versions and any other important settings (projects that need to be cached, etc. etc.). So we don't have to remember to manually keep them in sync with the yaml.

I'll be honest with you. I hate it. My understanding is that the only shared data between the yamls and sbt is the Scala versions. The disadvantages are that the yamls are huge, overly complicated, hard to read, have to be manually regenerated when we make changes, lock us into assumptions made upstreams (even the setup-scala version), we can't use dependabot to automatically propose action updates as they become available, the already terribly-messy `build.sbt` now becomes even more messy. All this to share Scala versions? That's my current impression.

In light of all of those concerns, do you still think it's a good idea to go ahead with this? If so, I'd love to hear why (genuinely, not antagonistically) and despite my scepticism I'm totally willing to trust in your judgement and give it a try. It wouldn't be the first time I don't understand the value of something to then come around after being forced to use it for a while :)

**Comment by armanbilge** [comment]
> I'll be honest with you. I hate it.

That's fine, let's dump it! Not gonna lie, it did not come out very nice in this situation. I see what you mean about upstream assumptions, I think I personally appreciate all the machinery it sets up in my own projects (caching, streamlined publishing, etc.). But maybe not appropriate for this project.

**Comment by japgolly** [comment]
Ok np but only if you genuinely agree. Like I said I'm totally open to giving it a try and seeing how it goes, and I'm honestly hoping to learn of some cool aspect that I'm just not seeing atm. If I'm wrong I want to leeeeeeeeeaaaaaaaaaaaaaaarrrrrrrrrrrnnnnnnnnnn :smile:

**Comment by armanbilge** [comment]
Maybe when we get around to 2.0, I can remake this PR (it will be much nicer!!) and you can see if you like it any better 😁 

> The disadvantages are that the yamls are huge, overly complicated, hard to read, have to be manually regenerated when we make changes, lock us into assumptions made upstreams

For me, these are what I consider the advantages! I don't think about the yaml anymore, I just appreciate all the machinery that's been installed for me automatically lol!

**Comment by japgolly** [comment]
>  I just appreciate all the machinery that's been installed for me automatically lol!

That's interesting cos I'm looking at all the new sbt config and thinking the same haha. Maybe it's just what we're used to. If I was already used to that sbt config and hadn't read up on the gha yml, and you had already been working with the yamls and had never seen the sbt way before, we might be saying the same things but on the opposite sides, haha. But ok cool, we let's reconsider over in 2.x. Thanks for the effort and taking the time to have this discussion with me hey. I'm sorry we're ending up closing this and appreciate the work you did.

**Pull request closed without merging**